File: python/toko_elektronik.py
class TokoElektronik:
    # ===== constructor =====
    def __init__(self, id=0, nama="", harga=0, stok=0):
        self.__id = id
        self.__nama = nama
        self.__harga = harga
        self.__stok = stok
        self.__daftar_produk = []  # list untuk menyimpan produk

    # ===== Getter =====
    def get_id(self):
        return self.__id

    def get_nama(self):
        return self.__nama

    def get_harga(self):
        return self.__harga

    def get_stok(self):
        return self.__stok

    # ===== Setter =====
    def set_id(self, id):
        self.__id = id

    def set_nama(self, nama):
        self.__nama = nama

    def set_harga(self, harga):
        self.__harga = harga

    def set_stok(self, stok):
        self.__stok = stok

    # ===== Method =====
    # SEARCH
    # method untuk mencari produk berdasarkan id
    def cari_produk(self, id):
        i = 0
        while i < len(self.__daftar_produk):
            if self.__daftar_produk[i].get_id() == id:
                return self.__daftar_produk[i]
            i += 1
        return None
    
    # method cari produk berdasarkan keyword nama
    def cari_produk_by_nama(self, keyword):
        hasil = []
        i = 0
        while i < len(self.__daftar_produk):
            if keyword.lower() in self.__daftar_produk[i].get_nama().lower():
                hasil.append(self.__daftar_produk[i])
            i += 1
        return hasil



    # CREATE
    # method untuk menambahkan produk ke list
    def tambah_produk(self, produk):
        self.__daftar_produk.append(produk)
        print("Produk berhasil ditambahkan!")

    # READ
    # method untuk menampilkan produk
    def tampilkan_produk(self):
        if len(self.__daftar_produk) == 0:
            print("Belum ada produk.")
        else:
            print("\nDaftar Produk:")
            for p in self.__daftar_produk:
                print(f"ID: {p.get_id()} | Nama: {p.get_nama()} | Harga: {p.get_harga()} | Stok: {p.get_stok()}")

    # UPDATE
    # method untuk mengubah produk di daftarProduk berdasarkan id
    def update_produk(self, id):
        p = self.cari_produk(id)
        if p is None:
            print(f"Produk dengan ID {id} tidak ditemukan.")
        else:
            nama = input("Nama baru: ")
            harga = int(input("Harga baru: "))
            stok = int(input("Stok baru: "))
            p.set_nama(nama)
            p.set_harga(harga)
            p.set_stok(stok)
            print("Produk berhasil diubah!")

    # DELETE
    # method untuk menghapus produk di daftarProduk
    def hapus_produk(self, id):
        i = 0
        while i < len(self.__daftar_produk):
            if self.__daftar_produk[i].get_id() == id:
                self.__daftar_produk.pop(i)
                print(f"Produk dengan ID {id} berhasil dihapus.")
                return
            i += 1
        print(f"Produk dengan ID {id} tidak ditemukan.")

File: cpp/TokoElektronik.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

// membuat class
class TokoElektronik {
    private:
        int id;
        string nama;
        int harga;
        int stok;

        // vector untuk menyimpan class TokoElektronik
        vector<TokoElektronik> daftarProduk;

    public:
        // constructor
        TokoElektronik(){
        }
        TokoElektronik(int id, string nama, int harga, int stok) {
            this->id = id;
            this->nama = nama;
            this->harga = harga;
            this->stok = stok;
        }

        // ===== getter ===== //
        int getId(){
            return this->id;
        }

        string getNama(){
            return this->nama;
        }

        int getHarga(){
            return this->harga;
        }

        int getStok(){
            return this->stok;
        }

        // ===== setter ===== //
        void setId(int id){
            this->id = id;
        }

        void setNama(string nama){
            this->nama = nama;
        }

        void setHarga(int harga){
            this->harga = harga;
        }

        void setStok(int stok){
            this->stok = stok;
        }

        // === method === //

        // SEARCH
        // method untuk mencari produk berdasarkan id
        TokoElektronik* cariProduk(int id) {
            int i = 0;
            while (i < daftarProduk.size()) {
                if (daftarProduk[i].getId() == id) {
                    return &daftarProduk[i]; // return alamat produk
                }
                i++;
            }
            return nullptr; // tidak ketemu
        }

        // method cari produk berdasarkan keyword nama
        vector<TokoElektronik> cariProdukByNama(const string& keyword) { // data disimpan ke vector
            vector<TokoElektronik> hasil;
            int i = 0;
            while (i < daftarProduk.size()) {
                if (daftarProduk[i].getNama().find(keyword) != string::npos) {
                    hasil.push_back(daftarProduk[i]);
                }
                i++;
            }
            return hasil;
        }


        // CREATE
        // method untuk menambahkan produk ke vector
        void tambahProduk(TokoElektronik &p){
            daftarProduk.push_back(p);
            cout << "Produk berhasil ditambahkan!\n";
        }

        // READ
        // method untuk menampilkan produk
        void tampilkanProduk(){
            if(daftarProduk.empty()){ // jika vector daftarProduk kosong
                cout << "Belum ada produk.\n";
            }else{ // jika vector daftarProduk tidak kosong
                cout << "\nDaftar Produk:\n";
                // menampilkan isi vector daftarProduk
                for (auto &p : daftarProduk) {
                    cout << "ID: " << p.getId()
                        << " | Nama: " << p.getNama()
                        << " | Harga: " << p.getHarga()
                        << " | Stok: " << p.getStok() << endl;
                }
            }
        }

        // UPDATE
        // method untuk mengubah produk di daftarProduk berdasarkan id
        void updateProduk(int id) {
            TokoElektronik* p = cariProduk(id); // cari produk dulu
            if (p == nullptr) {
                cout << "Produk dengan ID " << id << " tidak ditemukan.\n";
            }else{
                // kalau produk ada, baru minta input
                string nama;
                int harga, stok;
    
                cin.ignore(); // buang newline
                cout << "Nama baru: "; getline(cin, nama);
                cout << "Harga baru: "; cin >> harga;
                cout << "Stok baru: "; cin >> stok;
    
                // update field
                p->setNama(nama);
                p->setHarga(harga);
                p->setStok(stok);
    
                cout << "Produk berhasil diubah!\n";
            }
        }

        
        // DELETE
        // method untuk menghapus produk di daftarProduk
        void hapusProduk(int id){
            int i = 0;
            while (i < daftarProduk.size()) {
                if (daftarProduk[i].getId() == id) {
                    daftarProduk.erase(daftarProduk.begin() + i);
                    cout << "Produk dengan ID " << id << " berhasil dihapus.\n";
                    return;
                }
                i++;
            }
            cout << "Produk dengan ID " << id << " tidak ditemukan.\n";
        }

        // === destructor === //
        ~TokoElektronik() {
        }
};